src/table/data-grid-interaction-model.class.ts

```diff
@@ -0,0 +1,263 @@
+import {
+	BehaviorSubject,
+	Observable,
+	combineLatest
+} from "rxjs";
+import { map } from "rxjs/operators";
+import { TableAdapter } from "./table-adapter.class";
+import { tabbableSelectorIgnoreTabIndex, getFocusElementList } from "./../common/tab.service";
+
+/**
+ * The current and previous position in the grid.
+ *
+ * `current` and `previous` are tuples that follow the `[row, column]` convention.
+ */
+export interface DataGridPosition {
+	current: [number, number];
+	previous: [number, number];
+}
+
+/**
+ * `DataGridInteractionModel` provides centralized control over arbitrary 2d grids, following the w3 specs.
+ *
+ * Refs:
+ *  - https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html
+ *  - https://www.w3.org/TR/wai-aria-practices/#grid
+ *
+ * Example usage (taken from `table.component`):
+ * ```
+// a standard HTML table
+const table = this.elementRef.nativeElement.querySelector("table") as HTMLTableElement;
+
+// `TableDomAdapter` implements `TableAdapter` and provides a consistent interface to query rows and columns in a table
+const tableAdapter = new TableDomAdapter(table);
+
+// the keydown events that we'll use for keyboard navigation of the table
+const keydownEventStream = fromEvent<KeyboardEvent>(table, "keydown");
+
+// the click events we'll use to ensure focus is updated correctly on click
+const clickEventStream = fromEvent<MouseEvent>(table, "click");
+
+// the `DataGridInteractionModel` instance!
+this.interactionModel = new DataGridInteractionModel(keydownEventStream, clickEventStream, tableAdapter);
+
+// subscribe to the combined position updates
+this.interactionModel.position.subscribe(event => {
+	const [currentRow, currentColumn] = event.current;
+	const [previousRow, previousColumn] = event.previous;
+
+	// query the TableAdapter for the cell at the current row and column ...
+	const currentElement = tableAdapter.getCell(currentRow, currentColumn);
+	// ... and make it focusable it
+	Table.setTabIndex(currentElement, 0);
+
+	// if the model has just initialized don't focus or reset anything
+	if (previousRow === -1 || previousColumn === -1) { return; }
+
+	// query the TableAdapter for the cell at the previous row and column ...
+	const previousElement = tableAdapter.getCell(previousRow, previousColumn);
+	// ... and make it unfocusable (now there is only a single focusable cell)
+	Table.setTabIndex(previousElement, -1);
+
+	// finally, focus the current cell (skipped during initilzation)
+	Table.focus(currentElement);
+});
+ * ```
```

Comment:
Probably meant it without the `*`. Gods of markdown parsing won't let me suggest that change.

Reply:
Hah! yeah, I'll have to run the docgen locally and check